calico_node/startup/startup.go

```diff
@@ -906,6 +919,69 @@ func ensureDefaultConfig(ctx context.Context, cfg *apiconfig.CalicoAPIConfig, c
 	return nil
 }
 
+// ensureMigrated ensures any data migration needed is done.
+func ensureMigrated(ctx context.Context, cfg *apiconfig.CalicoAPIConfig, c client.Interface) error {
+	_, err := c.ClusterInformation().Get(ctx, "default", options.GetOptions{})
+	if err == nil {
+		// Nothing to do ClusterInformation exists so data has already been migrated
+		return nil
+	} else if _, ok := err.(cerrors.ErrorResourceDoesNotExist); !ok {
+		// The error indicates a problem with accessing the resource
+		return err
+	} else {
+		_, cv1, err := clients.LoadClients("", "")
+		if err != nil {
+			return err
+		}
+		// Grab the version from the clientv1
+		v, err := getV2ClusterVersion(cv1)
+		if err != nil {
+			if _, ok := err.(cerrors.ErrorResourceDoesNotExist); ok {
+				return nil
+			} else {
+				return err
+			}
+		}
+
+		// Migrate only if it is possible to migrate from the current version
+		if !canMigrate(v) {
+			return errors.New(fmt.Sprintf("Migration to v3 requires a base of Calico v2.6.4+, currently at %s", v))
+		}
+		_, r := migrate.Migrate(c, cv1, false) // Do migration
+		if r != migrate.ResultOK {
+			return errors.New("Migration failed")
+		}
+		return nil
+	}
+}
+
+// getV2ClusterVersion reads the CalicoVersion from the v1 client interface.
+func getV2ClusterVersion(cv1 clients.V1ClientInterface) (string, error) {
```

Comment:
Very confusing to have a v2-named function taking a v1 client.  Suggest using v1 consistently to refer to the backlevel API.

Reply:
Good point.